File: lessons/lesson10/NotificationsTest/NotificationsTest/Government.m
//
//  Government.m
//  NotificationsTest
//
//  Created by Руслан on 24.09.16.
//  Copyright © 2016 Ruslan Timchenko. All rights reserved.
//

#import "Government.h"

NSString *const GovernmentTaxLevelDidChangeNotification = @"GovernmentTaxLevelDidChangeNotification";
NSString *const GovernmentSalaryDidChangeNotification = @"GovernmentSalaryDidChangeNotification";
NSString *const GovernmentPensionDidChangeNotification = @"GovernmentPensionDidChangeNotification";
NSString *const GovernmentAveragePriceDidChangeNotification = @"GovernmentAveragePriceDidChangeNotification";

NSString *const GovernmentTaxLevelUserInfoKey = @"GovernmentTaxLevelUserInfoKey";
NSString *const GovernmentSalaryUserInfoKey = @"GovernmentSalaryUserInfoKey";
NSString *const GovernmentPensionUserInfoKey = @"GovernmentPensionUserInfoKey";
NSString *const GovernmentAveragePriceUserInfoKey = @"GovernmentAveragePriceUserInfoKey";

@implementation Government

- (instancetype)init
{
    self = [super init];
    if (self) {
        self.taxLevel = 5.f;
        self.salary = 1000.f;
        self.pension = 500.f;
        self.averagePrice = 10.f;

    }
    return self;
}

- (void) setTaxLevel:(Float64)taxLevel {
    _taxLevel = taxLevel;
    
    NSDictionary *dictionary = @{ GovernmentTaxLevelUserInfoKey: [NSNumber numberWithFloat:taxLevel] };
    
    [[NSNotificationCenter defaultCenter] postNotificationName:GovernmentTaxLevelDidChangeNotification
                                                        object:nil
                                                      userInfo:dictionary];
}

- (void) setSalary:(Float64)salary {
    _salary = salary;
    
    NSDictionary *dictionary = @{ GovernmentSalaryUserInfoKey: [NSNumber numberWithFloat:salary] };
    
    [[NSNotificationCenter defaultCenter] postNotificationName:GovernmentSalaryDidChangeNotification
                                                        object:nil
                                                      userInfo:dictionary];
}

- (void) setPension:(Float64)pension {
    _pension = pension;
    
    NSDictionary *dictionary = @{ GovernmentPensionUserInfoKey: [NSNumber numberWithFloat:pension] };
    
    [[NSNotificationCenter defaultCenter] postNotificationName:GovernmentPensionDidChangeNotification
                                                        object:nil
                                                      userInfo:dictionary];
}

- (void) setAveragePrice:(Float64)averagePrice {
    _averagePrice = averagePrice;
    
    NSDictionary *dictionary = @{ GovernmentAveragePriceUserInfoKey: [NSNumber numberWithFloat:averagePrice] };
    
    [[NSNotificationCenter defaultCenter] postNotificationName:GovernmentAveragePriceDidChangeNotification
                                                        object:nil
                                                      userInfo:dictionary];
}

@end

File: Homework/Swift/HW10/HW10/Government.swift
//
//  Government.swift
//  HW10
//
//  Created by Руслан on 25.09.16.
//  Copyright © 2016 Ruslan Timchenko. All rights reserved.
//

import UIKit

let GovernmentTaxLevelDidChangeNotification = Notification.Name(rawValue: "GovernmentTaxLevelDidChangeNotification")
let GovernmentSalaryDidChangeNotification = Notification.Name(rawValue: "GovernmentSalaryDidChangeNotification")
let GovernmentPensionDidChangeNotification = Notification.Name(rawValue: "GovernmentPensionDidChangeNotification")
let GovernmentAveragePriceDidChangeNotification = Notification.Name(rawValue: "GovernmentAveragePriceDidChangeNotification")


let GovernmentTaxLevelUserInfoKey = "GovernmentTaxLevelUserInfoKey"
let GovernmentSalaryUserInfoKey = "GovernmentSalaryUserInfoKey"
let GovernmentPensionUserInfoKey = "GovernmentPensionUserInfoKey"
let GovernmentAveragePriceUserInfoKey = "GovernmentAveragePriceUserInfoKey"

class Government: NSObject {
    
// MARK: - Properties and Notifications
    
    var taxLevel: Float {
        didSet {
            let taxLevelDictionary = [GovernmentTaxLevelUserInfoKey: taxLevel]
            
            NotificationCenter.default.post(name: GovernmentTaxLevelDidChangeNotification,
                                            object: nil,
                                            userInfo: taxLevelDictionary)
        }
    }
    var salary: Float {
        didSet {
            let salaryDictionary = [GovernmentSalaryUserInfoKey: salary]
            
            NotificationCenter.default.post(name: GovernmentSalaryDidChangeNotification,
                                            object: nil,
                                            userInfo: salaryDictionary)
        }
    }
    var pension: Float {
        didSet {
            let pensionDictionary = [GovernmentPensionUserInfoKey: pension]
            
            NotificationCenter.default.post(name: GovernmentPensionDidChangeNotification,
                                            object: nil,
                                            userInfo: pensionDictionary)
        }
    }
    var averagePrice: Float {
        didSet {
            let averagePriceDictionary = [GovernmentAveragePriceUserInfoKey: averagePrice]
            
            NotificationCenter.default.post(name: GovernmentAveragePriceDidChangeNotification,
                                            object: nil,
                                            userInfo: averagePriceDictionary)
        }
    }
    
// MARK: - Initialization
    
    override init() {
        self.taxLevel = 13
        self.salary = 500
        self.pension = 200
        self.averagePrice = 10
        super.init()
        
        NotificationCenter.default.addObserver(self, selector: #selector(didEnterBackgroundNotification),
                                               name: NSNotification.Name.UIApplicationDidEnterBackground, object: nil)
        
        NotificationCenter.default.addObserver(self, selector: #selector(willEnterForegroundNotification),
                                               name: NSNotification.Name.UIApplicationWillEnterForeground, object: nil)
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
    
// MARK: - Notifications
    
    func didEnterBackgroundNotification(notification: Notification) {
        print("Government goes to sleep")
    }
    
    func willEnterForegroundNotification(notification: Notification) {
        print("Government wakes up")
    }


}
